File: packages/pneuma-knowledge-core/src/pneuma_knowledge_core/compile/supersession.py
"""Claim supersession: the mechanical form of "the world changed".

WHY THIS EXISTS
---------------
The two daily write tools express two things and conflate a third. `append_block` adds a
claim; `edit_claim` rewrites one in place. Neither says that a NEW claim replaces an OLD
one as the current state of the same fact — a promotion, a change of employer, a moved
deadline. Written with `edit_claim`, the old state vanishes from the readable page (it
survives only in git); written with `append_block`, two states sit side by side with no
relation, and "which one holds now" becomes a reading exercise. The compile contract used
to ask the model to "keep the old state, the time of change and the source" — prompt copy,
which the project's own discipline forbids as the only enforcement.

`supersede_claim` is the third verb. The old claim stays byte-for-byte; the new claim gets
a system-assigned anchor and carries a second HTML-comment marker naming what it replaces:

    - X 是恒印印刷对接人 [cite: s01 ¶8-9] <!-- c:a1f3 -->
    - X 自 2026-05 起任新华印务采购总监 [cite: s02 ¶3] <!-- c:c07e --> <!-- supersedes: c:a1f3 -->

The anchor grammar (`domain.ids.ANCHOR_MARK_RE`) is untouched — the supersedes marker is a
separate comment that the anchor regex does not match, so every existing reader of anchors
keeps working and invariant I4 keeps its one parser per syntax.

WHAT IS DERIVED FROM THE MARKER
-------------------------------
- the CURRENT view of a document: its claims with no successor;
- the HISTORY of one fact: the chain root → … → head;
- an as-of answer: walk the chain by each claim's cited-source date (no new field);
- a `claim_superseded` compile event (transitions.py), so the brief narrates a state
  change rather than "one claim added".

WHAT THE GATE ENFORCES (compile.gate.check_supersession)
--------------------------------------------------------
target exists; no self-supersession; at most one successor per claim (chains are linear);
no cycles; a superseded claim is frozen (its text may not change afterwards); the
superseding claim carries new evidence (a `[cite:]` marker of its own). "Only new evidence
may supersede an old state" is thereby a write-time rejection, not a reminder.

Everything here is pure and sync.
"""

from __future__ import annotations

import re
from collections.abc import Iterable, Mapping

from ..domain.ids import ANCHOR_MARK_RE, extract_anchors
from .anchor_ops import anchored_blocks

#: The supersedes marker. Deliberately NOT a variant of the anchor comment: `<!-- c:… -->`
#: stays the one and only anchor syntax; this comment names a predecessor.
SUPERSEDES_MARK_RE = re.compile(r"<!--\s*supersedes:\s*c:([0-9a-f]{4,})\s*-->")


def supersedes_marker(old_anchor: str) -> str:
    """The marker text a superseding claim carries (single spelling, written here only)."""
    return f"<!-- supersedes: c:{old_anchor} -->"


def block_anchor(block: str) -> str | None:
    """The anchor a claim block carries as its own identity: the LAST anchor mark in the
    block (the write tools put a block's anchor on its final line)."""
    marks = ANCHOR_MARK_RE.findall(block)
    return marks[-1] if marks else None


def block_supersedes(block: str) -> list[str]:
    """Every predecessor anchor a block names (the gate requires exactly one)."""
    return SUPERSEDES_MARK_RE.findall(block)


def supersessions(body: str) -> dict[str, str]:
    """new anchor → old anchor, for every superseding claim in `body`.

    A block naming several predecessors is reported under its first one here; the gate
    rejects such a block, so committed canonical never contains one.
    """
    out: dict[str, str] = {}
    for block in anchored_blocks(body):
        olds = block_supersedes(block)
        if not olds:
            continue
        new = block_anchor(block)
        if new is not None:
            out[new] = olds[0]
    return out


def superseded_index(bodies: Mapping[str, str]) -> dict[str, tuple[str, str]]:
    """old anchor → (path of the successor, successor anchor), repository-wide."""
    out: dict[str, tuple[str, str]] = {}
    for path, body in bodies.items():
        for new, old in supersessions(body).items():
            out[old] = (path, new)
    return out


def current_blocks(body: str, superseded: Iterable[str]) -> list[str]:
    """The claims of `body` that hold NOW: anchored blocks with no successor anywhere.

    `superseded` is the repository-wide set of replaced anchors (`superseded_index` keys):
    a successor may live in another document (e.g. the active page superseding a claim
    archived in a frozen volume), so the current view of one page is a repo-level fact.
    """
    dead = set(superseded)
    return [b for b in anchored_blocks(body) if block_anchor(b) not in dead]


def chains(bodies: Mapping[str, str]) -> list[list[str]]:
    """Every supersession chain as an ordered anchor list, root first, head last.

    A chain root is a claim that supersedes nothing; the walk follows old → new. Cycles
    cannot occur in committed canonical (gate), but the walk still refuses to loop so a
    hand-edited repository cannot hang a reader.
    """
    successor: dict[str, str] = {}
    predecessor_of: set[str] = set()
    for body in bodies.values():
        for new, old in supersessions(body).items():
            successor[old] = new
            predecessor_of.add(new)
    result: list[list[str]] = []
    for root in sorted(a for a in successor if a not in predecessor_of):
        chain = [root]
        seen = {root}
        cursor = root
        while cursor in successor and successor[cursor] not in seen:
            cursor = successor[cursor]
            chain.append(cursor)
            seen.add(cursor)
        result.append(chain)
    return result


def block_by_anchor(bodies: Mapping[str, str]) -> dict[str, tuple[str, str]]:
    """anchor → (path, block text), repository-wide — the lookup chains resolve through."""
    out: dict[str, tuple[str, str]] = {}
    for path, body in bodies.items():
        for block in anchored_blocks(body):
            anchor = block_anchor(block)
            if anchor is not None:
                out[anchor] = (path, block)
    return out


__all__ = [
    "SUPERSEDES_MARK_RE",
    "block_anchor",
    "block_by_anchor",
    "block_supersedes",
    "chains",
    "current_blocks",
    "superseded_index",
    "supersedes_marker",
    "supersessions",
]

File: apps/web/src/lib/supersession.ts
/**
 * Claim supersession, read client-side off the canonical projection.
 *
 * A superseding claim carries a second HTML comment after its own anchor comment:
 *
 *     - X 自 2026-05 起任采购总监 [cite: s02 ¶3] <!-- c:c07e --> <!-- supersedes: c:a1f3 -->
 *
 * Semantics (`compile/supersession.py`): the new claim REPLACES `a1f3` as the current state
 * of that fact; `a1f3` stays in the document byte-for-byte as frozen history. The marker is
 * deliberately not an anchor variant — `<!-- c:… -->` remains the one anchor syntax — so the
 * two regexes here can never match each other's comment.
 *
 * The projection ships no supersession field, and it needs none: the relation is in the body
 * the exporter already sends. Two facts are repository-wide rather than per-document, so the
 * index is built over ALL documents:
 *
 * - whether a claim is superseded (its successor may live in another document — an active
 *   page superseding a state archived in a frozen volume);
 * - where the successor is, so a chip can jump to it.
 *
 * Pure and dependency-free: this module imports nothing at runtime, so it is transpiled and
 * tested standalone.
 */

import type { Claim, DocumentRecord } from "./types";

/** The supersession marker. One spelling, declared once; claim.ts strips prose with it. */
export const SUPERSEDES_COMMENT_RE = /<!--\s*supersedes:\s*c:([0-9a-f]{4,})\s*-->/i;

/** Where one claim lives — enough to select it across views. */
export interface ClaimSite {
  path: string;
  documentId: string | null;
  anchor: string;
}

export interface SupersessionIndex {
  /** superseding anchor → the anchor it replaces */
  supersedes: Map<string, string>;
  /** superseded anchor → where its successor lives */
  successorOf: Map<string, ClaimSite>;
  /** anchor → where that claim lives (both ends of every chip resolve through this) */
  siteOf: Map<string, ClaimSite>;
}

/** The audit text of a claim: `raw_text` when the exporter ships one, else `text`. */
function claimSource(claim: Claim): string {
  return claim.raw_text ?? claim.text ?? "";
}

/** The anchor a claim block replaces, or null. */
export function supersededAnchor(text: string): string | null {
  const m = SUPERSEDES_COMMENT_RE.exec(text);
  return m ? m[1].toLowerCase() : null;
}

export function emptySupersessionIndex(): SupersessionIndex {
  return { supersedes: new Map(), successorOf: new Map(), siteOf: new Map() };
}

/**
 * Build the repository-wide index. A claim naming several predecessors (which the compile
 * gate rejects, so committed canonical never holds one) is read by its first marker, and a
 * second successor for the same predecessor is ignored — chains stay linear here even if a
 * hand-edited repository is not.
 */
export function buildSupersessionIndex(docs: DocumentRecord[]): SupersessionIndex {
  const index = emptySupersessionIndex();
  for (const doc of docs ?? []) {
    for (const claim of doc.claims ?? []) {
      const anchor = claim.anchor?.toLowerCase();
      if (!anchor) continue;
      const site: ClaimSite = {
        path: doc.path,
        documentId: doc.document_id ?? null,
        anchor,
      };
      if (!index.siteOf.has(anchor)) index.siteOf.set(anchor, site);
      const old = supersededAnchor(claimSource(claim));
      // A claim superseding itself is a broken write, not a chain of one.
      if (!old || old === anchor) continue;
      if (!index.supersedes.has(anchor)) index.supersedes.set(anchor, old);
      if (!index.successorOf.has(old)) index.successorOf.set(old, site);
    }
  }
  return index;
}

/** True when some other claim has replaced this one as the current state. */
export function isSuperseded(index: SupersessionIndex, anchor: string | null): boolean {
  return !!anchor && index.successorOf.has(anchor.toLowerCase());
}

/** The claim this one replaced, if any. */
export function supersededBy(
  index: SupersessionIndex,
  anchor: string | null,
): string | null {
  return (anchor && index.supersedes.get(anchor.toLowerCase())) || null;
}

/** How many of a document's claims are frozen history. */
export function supersededCount(doc: DocumentRecord, index: SupersessionIndex): number {
  let n = 0;
  for (const claim of doc.claims ?? []) if (isSuperseded(index, claim.anchor)) n += 1;
  return n;
}

/** True when this document takes part in a chain at all — as history or as a successor. */
export function documentHasSupersession(
  doc: DocumentRecord,
  index: SupersessionIndex,
): boolean {
  for (const claim of doc.claims ?? []) {
    const anchor = claim.anchor?.toLowerCase();
    if (!anchor) continue;
    if (index.successorOf.has(anchor) || index.supersedes.has(anchor)) return true;
  }
  return false;
}

/** The claims of a document that hold NOW: those nothing has replaced. */
export function currentClaims(doc: DocumentRecord, index: SupersessionIndex): Claim[] {
  return (doc.claims ?? []).filter((c) => !isSuperseded(index, c.anchor));
}

/**
 * The whole chain one anchor belongs to, oldest first. The walk refuses to loop, so a
 * hand-edited repository with a cycle cannot hang a reader.
 */
export function supersessionChain(
  index: SupersessionIndex,
  anchor: string | null,
): string[] {
  if (!anchor) return [];
  const start = anchor.toLowerCase();
  const back: string[] = [start];
  const seen = new Set([start]);
  let cursor = start;
  for (;;) {
    const older = index.supersedes.get(cursor);
    if (!older || seen.has(older)) break;
    back.unshift(older);
    seen.add(older);
    cursor = older;
  }
  cursor = start;
  for (;;) {
    const newer = index.successorOf.get(cursor)?.anchor;
    if (!newer || seen.has(newer)) break;
    back.push(newer);
    seen.add(newer);
    cursor = newer;
  }
  return back;
}
